crates/trigger/src/loader.rs

```diff
@@ -13,16 +11,46 @@ use tokio::fs;
 
 use spin_common::{ui::quoted_path, url::parse_file_url};
 
+/// Compilation status of all components of a Spin application
+pub enum CompilationStatus {
+    #[cfg(feature = "aot-compilation")]
+    /// All components are componentized and ahead of time (AOT) compiled to cwasm.
+    AllAotComponents,
+    /// No components are ahead of time (AOT) compiled.
+    NoneAot,
+}
+
+/// Loader for the Spin runtime
 pub struct TriggerLoader {
+    /// Working directory where files for mounting exist.
     working_dir: PathBuf,
+    /// Set the static assets of the components in the temporary directory as writable.
     allow_transient_write: bool,
+    /// Declares the compilation status of all components of a Spin application.
+    compilation_status: CompilationStatus,
 }
 
 impl TriggerLoader {
     pub fn new(working_dir: impl Into<PathBuf>, allow_transient_write: bool) -> Self {
         Self {
             working_dir: working_dir.into(),
             allow_transient_write,
+            compilation_status: CompilationStatus::NoneAot,
+        }
+    }
+
+    /// Creates a TriggerLoader that loads AOT compiled components. It enables
```

Comment:
I think we'll want to spell out exactly what needs to happen here for this to be safe. i.e., that the `LockedComponentSource`  passed to `Loader::load_component` references a precompiled component compiled using the same version of `spin_core` used here and the same `Engine`.

Reply:
The most important point to get across is that it is _completely unsafe for untrusted input_.

Technically it is "safe" for almost any non-malicious input; it will just fail to deserialize unless compiled with a compatible engine.

Reply:
I updated the comment a little, but am still only mentioning that the requirement is `wasmtime` compilation (nothing Spin specific). Also, now the function is marked `unsafe`. @rylev do these changes look appropriate?

Reply:
I don't think marking the constructor as `unsafe` is the right approach. The contract with `unsafe` is that if the caller of the function can verify certain invariants are met *at the time of the function call* then the call is safe. However, it's the call to `Loader::load_component` that needs to verify the invariant (in this case that the component source is valid to be used with `Component::deserialize`). 

`unsafe` is IMO inappropriate because the invariant that needs to be satisfied cannot be satisfied in code. It's up to whoever is running the software to ensure the invariant holds. This is why I think a feature flag that prominently displays the safety risks is the right approach.

Reply:
I now see @lann's comment below pointing to similar usage in `cranelift`. However, I think this is categorically different usage. 

The usage in the `cranelift` API can still be verified as correct by users by verifying its usage is correct in combination with the rest of the usage of `cranelift` - in other words, the safety properties can be verified statically albeit manually. The usage here on the other hand, cannot be verified statically as correct - it depends on how the software is run - the same code can be correct when run in one context and incorrect when run in another. That is not what `unsafe` is meant to signal.

Reply:
> the safety properties can be verified statically albeit manually

By this definition I don't think `deserialize_file` would be `unsafe` either; I don't think any unsafety occurs until the component is instantiated and you certainly can't verify its safety statically...

Reply:
`deserialize_file` is different in that the safety check happens at the call site of the function marked `unsafe`. You should only call `deserialize_file` if you're sure the input *to that function* meets some invariants. That's different than calling a function and promising all future input into the system as a whole meets some invariants. 

Taking a step back... we both are trying to achieve the same thing here: ensuring that users of these APIs know what the right thing to do is. I personally think the feature flag is the right approach. It says "don't build and run this software unless you know what you're doing", and I think it does it in a better way than marking the function `unsafe`. If others disagree though that's fine, I won't try to block merging this PR.

Reply:
I would be fine with the feature flag only if "unsafe" appears in it, but I would personally keep this `unsafe` as well (belt _and_ suspenders), just so `unsafe` appears _somewhere_ in the calling code. :shrug:

Reply:
> just so unsafe appears somewhere in the calling code. 

I tend to agree with this and this was the intent originally.

Reply:
We are aiming to get this in tomorrow before we can cut a release. @rylev can we leave unsafe on the setter? @lann does this LGTY?

Reply:
I'm not willing to die on this hill 😉 so :shipit:. It's certainly better than *not* marking a function as `unsafe` when it clearly is.